doc/devguide/extending/app-layer/transactions.rst

```diff
@@ -0,0 +1,321 @@
+or ``ntp``, that will most
+likely happen once per transaction, by the time of its completion. In other cases, like with HTTP, this may happen at intermediary states.
+
+In ``OutputTxLog``, the engine will compare current state with the value defined for the logging to happen, per flow
+direction (``logger->tc_log_progress``, ``logger->ts_log_progress``). If state is less than that value, the engine skips to
+the next logger. Code snippet from: suricata/src/output-tx.c:
+
+.. code-block:: c
+
+    static TmEcode OutputTxLog(ThreadVars *tv, Packet *p, void *thread_data)
+    {
+        .
+        .
+        .
+            if ((ts_eof && tc_eof) || last_pseudo) {
+                SCLogDebug("EOF, so log now");
+            } else {
+                if (logger->LogCondition) {
+                    int r = logger->LogCondition(tv, p, alstate, tx, tx_id);
+                    if (r == FALSE) {
+                        SCLogDebug("conditions not met, not logging");
+                        goto next_logger;
+                    }
+                } else {
+                    if (tx_progress_tc < logger->tc_log_progress) {
+                        SCLogDebug("progress not far enough, not logging");
+                        goto next_logger;
+                    }
+
+                    if (tx_progress_ts < logger->ts_log_progress) {
+                        SCLogDebug("progress not far enough, not logging");
+                        goto next_logger;
+                    }
+                }
+             }
+        .
+        .
+        .
+    }
+
+Rule Matching
+~~~~~~~~~~~~~
+
+Transaction progress is also used for certain keywords to know what is the minimum state before we can expect a match: until that, Suricata won't even try to look for the patterns.
+
+As seen in ``DetectAppLayerMpmRegister2`` that has ``int progress`` as parameter, and ``DetectAppLayerInspectEngineRegister2``, which expects ``int tx_min_progress``, for instance. In the code snippet,
+``HTTP2StateDataClient``, ``HTTP2StateDataServer`` and ``0`` are the values passed to the functions.
+
+
+.. code-block:: c
+
+    void DetectFiledataRegister(void)
+    {
+        .
+        .
+        DetectAppLayerMpmRegister2("file_data", SIG_FLAG_TOSERVER, 2,
+                PrefilterMpmFiledataRegister, NULL,
+                ALPROTO_HTTP2, HTTP2StateDataClient);
+        DetectAppLayerMpmRegister2("file_data", SIG_FLAG_TOCLIENT, 2,
+                PrefilterMpmFiledataRegister, NULL,
+                ALPROTO_HTTP2, HTTP2StateDataServer);
+        .
+        .
+        DetectAppLayerInspectEngineRegister2("file_data",
+            ALPROTO_HTTP2, SIG_FLAG_TOCLIENT, HTTP2StateDataServer,
+            DetectEngineInspectFiledata, NULL);
+        DetectAppLayerInspectEngineRegister2(
+                "file_data", ALPROTO_FTPDATA, SIG_FLAG_TOSERVER, 0, DetectEngineInspectFiledata, NULL);
+        .
+        .
+    }
+
+Progress Tracking
```

Comment:
recent changes set an upper limit to the value of the process. It has to be <48. This is due to how the detection engine tracks detection across the states.